File: python_files/text_games/dungeon_delve/player.py
from abc import ABC

player_types: list[str] = [
    "warrior",
    "ranger",
    "mage",
    "cleric",
    "rogue",
    "paladin",
    "berserker",
    "necromancer",
    "druid",
    "monk",
]


class Player(ABC):
    def __init__(self) -> None:
        self.x: int = 0
        self.y: int = 0
        self.level: int = 1
        self.gold: int = 0

    def is_alive(self) -> bool:
        return self.base_health > 0

    def move_north(self) -> None:
        self.y -= 1

    def move_south(self) -> None:
        self.y += 1

    def move_east(self) -> None:
        self.x += 1

    def move_west(self) -> None:
        self.x -= 1


class Warrior(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 90
        self.base_damage: int = 8
        self.health_growth_rate: int = 1.12
        self.damage_growth_rate: int = 1.06


class Ranger(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 60
        self.base_damage: int = 11
        self.health_growth_rate: int = 1.08
        self.damage_growth_rate: int = 1.12


class Mage(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 50
        self.base_damage: int = 14
        self.health_growth_rate: int = 1.05
        self.damage_growth_rate: int = 1.15


class Cleric(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 65
        self.base_damage: int = 6
        self.health_growth_rate: int = 1.10
        self.damage_growth_rate: int = 1.04


class Rogue(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 55
        self.base_damage: int = 13
        self.health_growth_rate: int = 1.07
        self.damage_growth_rate: int = 1.14


class Paladin(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 80
        self.base_damage: int = 9
        self.health_growth_rate: int = 1.11
        self.damage_growth_rate: int = 1.08


class Berserker(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 70
        self.base_damage: int = 12
        self.health_growth_rate: int = 1.10
        self.damage_growth_rate: int = 1.13


class Necromancer(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 50
        self.base_damage: int = 10
        self.health_growth_rate: int = 1.06
        self.damage_growth_rate: int = 1.10


class Druid(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 65
        self.base_damage: int = 9
        self.health_growth_rate: int = 1.10
        self.damage_growth_rate: int = 1.09


class Monk(Player):
    def __init__(self) -> None:
        super().__init__()

        self.base_health: int = 60
        self.base_damage: int = 10
        self.health_growth_rate: int = 1.09
        self.damage_growth_rate: int = 1.10

File: c_files/text_games/dungeon_delve/src/player.c
#include "player.h"

#include "commands.h"

void player_init(Player *player, PlayerType type, Vector2 start_pos) {
  player->position.x = start_pos.x;
  player->position.y = start_pos.y;
  player->type = type;
  player->level = 1;
  player->gold = 0;

  if (player->type == WARRIOR) {
    player->base_health = 90;
    player->base_damage = 8;
    player->health_growth_rate = 1.12f;
    player->damage_growth_rate = 1.06f;
  }
  if (player->type == RANGER) {
    player->base_health = 60;
    player->base_damage = 11;
    player->health_growth_rate = 1.08f;
    player->damage_growth_rate = 1.12f;
  }
  if (player->type == MAGE) {
    player->base_health = 50;
    player->base_damage = 14;
    player->health_growth_rate = 1.05f;
    player->damage_growth_rate = 1.15f;
  }
  if (player->type == CLERIC) {
    player->base_health = 65;
    player->base_damage = 6;
    player->health_growth_rate = 1.10f;
    player->damage_growth_rate = 1.04f;
  }
  if (player->type == ROGUE) {
    player->base_health = 55;
    player->base_damage = 13;
    player->health_growth_rate = 1.07f;
    player->damage_growth_rate = 1.14f;
  }
  if (player->type == PALADIN) {
    player->base_health = 80;
    player->base_damage = 9;
    player->health_growth_rate = 1.11f;
    player->damage_growth_rate = 1.08f;
  }
  if (player->type == BERSERKER) {
    player->base_health = 70;
    player->base_damage = 12;
    player->health_growth_rate = 1.10f;
    player->damage_growth_rate = 1.13f;
  }
  if (player->type == NECROMANCER) {
    player->base_health = 50;
    player->base_damage = 10;
    player->health_growth_rate = 1.06f;
    player->damage_growth_rate = 1.10f;
  }
  if (player->type == DRUID) {
    player->base_health = 65;
    player->base_damage = 9;
    player->health_growth_rate = 1.10f;
    player->damage_growth_rate = 1.09f;
  }
  if (player->type == MONK) {
    player->base_health = 60;
    player->base_damage = 10;
    player->health_growth_rate = 1.09f;
    player->damage_growth_rate = 1.10f;
  }
}

bool player_is_alive(Player *player) { return player->base_health > 0; }

void player_move(Player *player, Commands command) {
  if (command == NORTH) {
    player->position.y -= 1;
  } else if (command == SOUTH) {
    player->position.y += 1;
  } else if (command == EAST) {
    player->position.x += 1;
  } else if (command == WEST) {
    player->position.x -= 1;
  }
}